src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs

```diff
@@ -930,8 +906,8 @@ public unsafe void SetLocalSignature(byte* localSignature, int signatureSize)
             if (signatureSize < 0)
                 throw new ArgumentOutOfRangeException(nameof(signatureSize), SR.ArgumentOutOfRange_GenericPositive);
 
-            if (signatureSize > 0 && localSignature == null)
-                throw new ArgumentNullException(nameof(localSignature));
+            if (signatureSize > 0)
+                ArgumentNullException.ThrowIfNull(localSignature);
```

Comment:
> Is this the proper exception to be thrown now?

It's the same as it was before.  Previously the code was:
```C#
if (signatureSize > 0 && localSignature == null)
    throw new ArgumentNullException(nameof(localSignature));
```
which is the same as:
```C#
if (signatureSize > 0)
    if (localSignature == null)
        throw new ArgumentNullException(nameof(localSignature));
```
which is the same as:
```C#
if (signatureSize > 0)
    ArgumentNullException.ThrowIfNull(localSignature));
```
which is what the new version is.

Reply:
I understand that it is technically the same logic and it could have been misleading before but I think with this change it could be doubly misleading.

Before if someone got an `ArgumentNullException` it would point them to `localSignature` but technically they could be mislead because the issue could ALSO be with `signatureSize`. Now, `!!` takes care of throwing if `localSignature` is null. So at the point of `if (signatureSize > 0)` we know for a fact that `localSignature` is not null but we the exception would point them to `localSignature`. 

It's not a regression as the same inputs give you the same outputs but as a user this now reads wrong to me and I think an `ArgumentException` would be better suited now.

Reply:
> It's not a regression as the same inputs give you the same outputs but as a user this now reads wrong to me and I think an ArgumentException would be better suited now.

There are many places where if the code were being written from scratch we would choose to use a different exception type than is currently being used.  But changing from a more derived type to one of its base types is a breaking change, and so we need a good reason to do so.

Reply:
Totally fair, have a great day! Looking forward to doing this in my code!